### PR TITLE
"jsdoc/require-property-description" rule and fixes to events.js

### PR DESCRIPTION
**Pull request by SnailBones**
Added the "jsdoc/require-property-description" linter rule and made changes to make it pass

Fixes in events.js:

-    Replaced repeated incorrect @property {...} data with @type
-    Fixed incorrect property with error event.
-    Elaborated on @property {string} when limited to a set of events.
-    Changed event descriptions erroneously referring to all events as "mouse events"

### CODE REVIEW

File: src/ui/events.js [hunk 1]
```diff
@@ -96,7 +96,7 @@ export class MapMouseEvent extends Event {
      */
     features: Array<Object> | void;
 
-    /**
+    /**.
```
Comment (SnailBones) [inline src/ui/events.js:99]:
```suggestion
    /**
```

File: src/ui/events.js [hunk 2]
```diff
@@ -14,7 +14,7 @@ import type LngLat from '../geo/lng_lat.js';
  * mouse events of specific types such as 'click' or 'hover'.
  * For a full list of available events, see [`Map` events](/mapbox-gl-js/api/map/#map-events).
  *
- * @extends {Object}
+ * @augments {object}
```
Comment (SnailBones) [inline src/ui/events.js:17]:
```suggestion
 * @extends {Object}
```

File: src/ui/events.js [hunk 3]
```diff
@@ -142,7 +142,7 @@ export class MapMouseEvent extends Event {
  * mouse events of specific types such as 'touchstart' or 'touchend'.
  * For a full list of available events, see [`Map` events](/mapbox-gl-js/api/map/#map-events).
  *
- * @extends {Object}
+ * @augments {object}
```
Comment (SnailBones) [inline src/ui/events.js:145]:
```suggestion
 * @extends {Object}
```

File: src/ui/events.js [hunk 4]
```diff
@@ -176,7 +176,7 @@ export class MapMouseEvent extends Event {
  * });
  * @see [`Map` events documentation](https://docs.mapbox.com/mapbox-gl-js/api/map/#map-events)
  * @see [Create a draggable point](https://docs.mapbox.com/mapbox-gl-js/example/drag-a-point/)
-*/
+ */
```
Comment (SnailBones) [inline src/ui/events.js:179]:
```suggestion
*/
```

File: src/ui/events.js [hunk 5]
```diff
@@ -242,7 +242,7 @@ export class MapTouchEvent extends Event {
      */
     features: Array<Object> | void;
 
-    /**
+    /**.
```
Comment (SnailBones) [inline src/ui/events.js:245]:
```suggestion
    /**
```

File: src/ui/events.js [hunk 6]
```diff
@@ -290,14 +290,14 @@ export class MapTouchEvent extends Event {
  * mouse events of specific types such as 'wheel'.
  * For a full list of available events, see [`Map` events](/mapbox-gl-js/api/map/#map-events).
  *
- * @extends {Object}
+ * @augments {object}
```
Comment (SnailBones) [inline src/ui/events.js:293]:
```suggestion
 * @extends {Object}
```

File: src/ui/events.js [hunk 7]
```diff
@@ -290,14 +290,14 @@ export class MapTouchEvent extends Event {
...
  * @example
  * // Example of a MapWheelEvent of type "wheel"
  * map.on('wheel', (e) => {
  *     console.log('event type:', e.type);
  *     // event type: wheel
  * });
-* @see [`Map` events documentation](https://docs.mapbox.com/mapbox-gl-js/api/map/#map-events)
+ * @see [`Map` events documentation](https://docs.mapbox.com/mapbox-gl-js/api/map/#map-events)
```
Comment (SnailBones) [inline src/ui/events.js:300]:
```suggestion
* @see [`Map` events documentation](https://docs.mapbox.com/mapbox-gl-js/api/map/#map-events)
```

File: src/ui/events.js [hunk 8]
```diff
@@ -319,6 +319,7 @@ export class MapWheelEvent extends Event {
      * Prevents subsequent default processing of the event by the map.
      *
      * Calling this method will prevent the the behavior of {@link ScrollZoomHandler}.
+     *
```
Comment (SnailBones) [inline src/ui/events.js:322]:
```suggestion
     * Calling this method will prevent the the behavior of {@link ScrollZoomHandler}.
```

File: src/style/style_image.js [hunk 9]
```diff
@@ -52,9 +52,9 @@ export function renderStyleImage(image: StyleImage) {
  * can be used to update the image.
  *
  * @interface StyleImageInterface
- * @property {number} width
- * @property {number} height
- * @property {Uint8Array | Uint8ClampedArray} data
+ * @property {number} width Width in pixels.
+ * @property {number} height Height in Pixels.
+ * @property {Uint8Array | Uint8ClampedArray} data Byte array representing the image. Size must be width × height × 4 (for each RGBA channel).
```
Comment (ryanhamley) [inline src/style/style_image.js:57]: nit: How about something like `(to ensure space for all four channels in an RGBA color).` to be more explicit?

File: src/ui/events.js [hunk 10]
```diff
@@ -350,13 +350,13 @@ export class MapWheelEvent extends Event {
 }
 
 /**
- * `MapBoxZoomEvent` is a class used by other classes to generate
- * mouse events of specific types such as 'boxzoomstart' or 'boxzoomend'.
+ * `MapBoxZoomEvent` is a class used to generate
+ * events 'boxzoomstart', 'boxzoomend' and 'boxzoomcancel'.
```
Comment (ryanhamley) [inline src/ui/events.js:354]: nit: `the events`

File: src/ui/events.js [hunk 11]
```diff
@@ -376,18 +376,18 @@ export type MapBoxZoomEvent = {
...
- * `MapDataEvent` is a class used by other classes to generate
- * mouse events of specific types such as 'sourcedata' or 'dataloading'.
+ * `MapDataEvent` is a class used to generate
+ * events related to loading data, styles and sources.
  * For a full list of available events, see [`Map` events](/mapbox-gl-js/api/map/#map-events).
  *
- * @typedef {Object} MapDataEvent
- * @property {string} type The type of originating event. For a full list of available events, see [`Map` events](/mapbox-gl-js/api/map/#map-events).
- * @property {string} dataType The type of data that has changed. One of `'source'` or `'style'`, where `'source'` refers to the data associated with any source, and `'style'` refers to the entire [style](https://docs.mapbox.com/help/glossary/style/) used by the map.
+ * @typedef {object} MapDataEvent
```
Comment (ryanhamley) [inline src/ui/events.js:383]: should be `Object`

File: src/ui/events.js [hunk 12]
```diff
@@ -376,18 +376,18 @@ export type MapBoxZoomEvent = {
...
+ * @property {('data' | 'dataloading' | 'styledata' | 'styledataloading' | 'sourcedata'| 'sourcedataloading')} type The type of originating event. For a full list of available events, see [`Map` events](/mapbox-gl-js/api/map/#map-events).
+ * @property {('source' | 'style')} dataType The type of data that has changed. One of `'source'` or `'style'`, where `'source'` refers to the data associated with any source, and `'style'` refers to the entire [style](https://docs.mapbox.com/help/glossary/style/) used by the map.
  * @property {boolean} [isSourceLoaded] True if the event has a `dataType` of `source` and the source has no outstanding network requests.
- * @property {Object} [source] The [style spec representation of the source](https://docs.mapbox.com/mapbox-gl-js/style-spec/sources/) if the event has a `dataType` of `source`.
+ * @property {object} [source] The [style spec representation of the source](https://docs.mapbox.com/mapbox-gl-js/style-spec/sources/) if the event has a `dataType` of `source`.
```
Comment (ryanhamley) [inline src/ui/events.js:387]: `Object`

File: src/ui/events.js [hunk 13]
```diff
@@ -376,18 +376,18 @@ export type MapBoxZoomEvent = {
...
  * @property {string} [sourceDataType] Included if the event has a `dataType` of `source` and the event signals
  * that internal data has been received or changed. Possible values are `metadata`, `content` and `visibility`.
- * @property {Object} [tile] The tile being loaded or changed, if the event has a `dataType` of `source` and
+ * @property {object} [tile] The tile being loaded or changed, if the event has a `dataType` of `source` and
```
Comment (ryanhamley) [inline src/ui/events.js:390]: `Object`

File: src/ui/events.js [hunk 14]
```diff
@@ -1101,8 +1102,8 @@ export type MapEvent =
      * @event boxzoomend
      * @memberof Map
      * @instance
-     * @type {Object}
-     * @property {MapBoxZoomEvent} data
+     * @type {object}
```
Comment (ryanhamley) [inline src/ui/events.js:1105]: `Object`

File: src/ui/events.js [hunk 15]
```diff
@@ -1191,7 +1192,7 @@ export type MapEvent =
      * @event load
      * @memberof Map
      * @instance
-     * @type {Object}
+     * @type {object}
```
Comment (ryanhamley) [inline src/ui/events.js:1195]: `Object`

**Review by ryanhamley — CHANGES_REQUESTED** [review]
Looks good overall. Fixing the `object` issues will fix the lint failures and I had a couple nits. How does the `@type` parameter show up in the docs UI?

File: src/style/style_image.js (hunk 9 above)
Comment (SnailBones) [reply src/style/style_image.js:57]:
```suggestion
 * @property {Uint8Array | Uint8ClampedArray} data Byte array representing the image. Size must be width × height × 4 (to ensure space for all four channels in an RGBA color)
```

File: src/ui/events.js (hunk 10 above)
Comment (SnailBones) [reply src/ui/events.js:354]:
```suggestion
 * the events 'boxzoomstart', 'boxzoomend' and 'boxzoomcancel'.
```

File: src/ui/events.js (hunk 11 above)
Comment (SnailBones) [inline src/ui/events.js:383]:
```suggestion
 * @typedef {Object} MapDataEvent
```

File: src/ui/events.js (hunk 12 above)
Comment (SnailBones) [inline src/ui/events.js:387]:
```suggestion
 * @property {Object} [source] The [style spec representation of the source](https://docs.mapbox.com/mapbox-gl-js/style-spec/sources/) if the event has a `dataType` of `source`.
```

File: src/ui/events.js (hunk 13 above)
Comment (SnailBones) [inline src/ui/events.js:390]:
```suggestion
 * @property {Object} [tile] The tile being loaded or changed, if the event has a `dataType` of `source` and
```

File: src/ui/events.js (hunk 14 above)
Comment (SnailBones) [reply src/ui/events.js:1105]:
```suggestion
     * @type {Object}
```

File: src/ui/events.js (hunk 15 above)
Comment (SnailBones) [reply src/ui/events.js:1195]:
```suggestion
     * @type {Object}
```

**Review by domlet — APPROVED** [review]
Thank you for your care in turning these from `@property` to `@type`! 💐

**Comment by domlet** [comment]
> How does the @type parameter show up in the docs UI? Can you include a screenshot of that?

Oh dear, thank you @ryanhamley for pushing on ^ this.

@SnailBones It looks to me like the result of switching from `@property` to `@type` is that the item is not rendered. Can we switch these back to `@property`?

## Screenshots

`publisher-production` | `aidan/require-description` |  `aidan/require-description`
--- | --- | ---
screenshot | screenshot | screenshot

**Review by domlet — CHANGES_REQUESTED** [review]
Looks like we should change a few of these back:
https://github.com/mapbox/mapbox-gl-js/pull/10870#issuecomment-881797089

**Comment by ryanhamley** [comment]
Ahhh yeah, this is a mixup. These events all have a property called `data` that is an instance of the event class that made the event, if that makes sense. So I think we need to revert the `@property` -> `@type` changes

**Comment by SnailBones** [comment]
I don't think they do have a `data` property. We had a discussion about this in [slack](https://mapbox.slack.com/archives/C024JPPM29Y/p1626291232090500) a few days ago.

From @arindam1993 :
> here the `data` thing is completely incorrect :sweat_smile: , there is no `e.data property`
The event itself is of type `MapMouseEvent`

**Comment by ryanhamley** [comment]
I missed that discussion! Sounds good to me as long as we can get `type` to render

File: src/ui/events.js [hunk 16]
```diff
@@ -813,7 +814,7 @@ export type MapEvent =
      * @event movestart
      * @memberof Map
      * @instance
-     * @property {{originalEvent: DragEvent}} data
+     * @type {{originalEvent: DragEvent}}
```
Comment (katydecorah) [inline src/ui/events.js:817]: @SnailBones I'm working on the UI for this update and noticed that this event type is formatted differently than others. Should this type be: 

```suggestion
     * @type {DragEvent}
```

File: src/style/style_image.js [hunk 17]
```diff
@@ -52,9 +52,9 @@ export function renderStyleImage(image: StyleImage) {
  * can be used to update the image.
  *
  * @interface StyleImageInterface
- * @property {number} width
- * @property {number} height
- * @property {Uint8Array | Uint8ClampedArray} data
+ * @property {number} width Width in pixels.
+ * @property {number} height Height in Pixels.
```
Comment (katydecorah) [inline src/style/style_image.js:56]: Nit: 
```suggestion
 * @property {number} height Height in pixels.
```

File: src/ui/events.js (hunk 16 above)
Comment (SnailBones) [reply src/ui/events.js]: Good catch! I was running Regex over them to convert them but it looks like I missed that one!

File: src/style/style_image.js [hunk 18]
```diff
@@ -52,9 +52,9 @@ export function renderStyleImage(image: StyleImage) {
  * can be used to update the image.
  *
  * @interface StyleImageInterface
- * @property {number} width
- * @property {number} height
- * @property {Uint8Array | Uint8ClampedArray} data
+ * @property {number} width Width in pixels.
```
Comment (rreusser) [inline src/style/style_image.js:55]: For readability, I like the [require-hyphen-before-param-description](https://github.com/gajus/eslint-plugin-jsdoc#eslint-plugin-jsdoc-rules-require-hyphen-before-param-description) rule, but it's tangential and not related to the PR under consideration.

```suggestion
 * @property {number} width - Width in pixels.
```

File: src/style/style_image.js [hunk 19]
```diff
@@ -52,9 +52,9 @@ export function renderStyleImage(image: StyleImage) {
...
+ * @property {number} height Height in pixels.
+ * @property {Uint8Array | Uint8ClampedArray} data Byte array representing the image. Size must be width × height × 4 (to ensure space for all four channels in an RGBA color)
```
Comment (rreusser) [inline src/style/style_image.js:57]:
```suggestion
 * @property {Uint8Array | Uint8ClampedArray} data Byte array representing the image. To ensure space for all four channels in an RGBA color, size must be width × height × 4.
```

File: src/ui/events.js [hunk 20]
```diff
@@ -357,13 +357,13 @@ export class MapWheelEvent extends Event {
 }
 
 /**
- * `MapBoxZoomEvent` is a class used by other classes to generate
- * mouse events of specific types such as 'boxzoomstart' or 'boxzoomend'.
+ * `MapBoxZoomEvent` is a class used to generate
+ * the events 'boxzoomstart', 'boxzoomend' and 'boxzoomcancel'.
```
Comment (rreusser) [inline src/ui/events.js:361]: Extreme nit: oxford comma 😬 

```suggestion
 * the events 'boxzoomstart', 'boxzoomend', and 'boxzoomcancel'.
```

File: src/ui/events.js [hunk 21]
```diff
@@ -390,13 +390,13 @@ export type MapBoxZoomEvent = {
 };
 
 /**
- * `MapDataEvent` is a class used by other classes to generate
- * mouse events of specific types such as 'sourcedata' or 'dataloading'.
+ * `MapDataEvent` is a class used to generate
+ * events related to loading data, styles and sources.
```
Comment (rreusser) [inline src/ui/events.js:394]: Nit:

```suggestion
 * events related to loading data, styles, and sources.
```

**Review by rreusser — APPROVED** [review]
A couple nits, one idea, and nothing blocking. 👍

File: src/style/style_image.js (hunk 18 above)
Comment (SnailBones) [reply src/style/style_image.js:55]: That does look nicer! Perhaps we should make a separate PR/issue?

**Review by katydecorah — APPROVED** [review]

**Pull request merged**